File: FlexTargetAndroid/app/src/main/java/com/flextarget/android/ui/ipsc/IpscSubmitViewModel.kt
package com.flextarget.android.ui.ipsc

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.flextarget.android.data.model.DrillRepeatSummary
import com.flextarget.android.data.model.ScoringUtility
import com.flextarget.android.data.remote.api.IpscMatch
import com.flextarget.android.data.remote.api.IpscScoreHits
import com.flextarget.android.data.remote.api.IpscScoreTargetRow
import com.flextarget.android.data.remote.api.IpscScorePenalties
import com.flextarget.android.data.remote.api.IpscScoreSubmitRequest
import com.flextarget.android.data.remote.api.IpscShooter
import com.flextarget.android.data.remote.api.IpscSquad
import com.flextarget.android.data.repository.IpscRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

/**
 * Steps in the IPSC submit dialog flow:
 *
 * Idle → LoadingMatches → MatchPicker
 *                              ↓ selectMatch
 *                        LoadingSquads → ShooterPicker
 *                                             ↓ selectShooter
 *                                        Confirm → Submitting → Success
 *                                                             → Error
 */
sealed class IpscSubmitStep {
    object Idle : IpscSubmitStep()
    object LoadingMatches : IpscSubmitStep()
    data class MatchPicker(val matches: List<IpscMatch>) : IpscSubmitStep()
    data class LoadingSquads(val matchId: Int) : IpscSubmitStep()
    data class ShooterPicker(val matchId: Int, val squads: List<IpscSquad>) : IpscSubmitStep()
    data class Confirm(val matchId: Int, val shooter: IpscShooter) : IpscSubmitStep()
    object Submitting : IpscSubmitStep()
    data class Success(val hitFactor: Double, val totalPoints: Int) : IpscSubmitStep()
    data class Error(val message: String, val retryAction: (() -> Unit)? = null) : IpscSubmitStep()
}

class IpscSubmitViewModel(
    private val repository: IpscRepository
) : ViewModel() {

    private val _step = MutableStateFlow<IpscSubmitStep>(IpscSubmitStep.Idle)
    val step: StateFlow<IpscSubmitStep> = _step

    // ---- Lifecycle ----

    /** Start the flow: fetch matches and move to MatchPicker. */
    fun start() {
        loadMatches(forceRefresh = false)
    }

    /** Go back one step (Confirm → ShooterPicker, ShooterPicker → MatchPicker). */
    fun back() {
        when (val current = _step.value) {
            is IpscSubmitStep.Confirm -> {
                val matchId = current.matchId
                viewModelScope.launch {
                    _step.value = IpscSubmitStep.LoadingSquads(matchId)
                    repository.getSquadQueue(matchId).fold(
                        onSuccess = { squads -> _step.value = IpscSubmitStep.ShooterPicker(matchId, squads) },
                        onFailure = { e ->
                            _step.value = IpscSubmitStep.Error(e.message ?: "Error loading squads") { selectMatch(matchId) }
                        }
                    )
                }
            }
            is IpscSubmitStep.ShooterPicker -> loadMatches(forceRefresh = false)
            else -> { /* no-op */ }
        }
    }

    /** Reset to Idle (call when dialog is dismissed). */
    fun dismiss() {
        _step.value = IpscSubmitStep.Idle
    }

    // ---- Step transitions ----

    fun refreshMatches() = loadMatches(forceRefresh = true)

    fun selectMatch(matchId: Int) {
        viewModelScope.launch {
            _step.value = IpscSubmitStep.LoadingSquads(matchId)
            repository.getSquadQueue(matchId).fold(
                onSuccess = { squads -> _step.value = IpscSubmitStep.ShooterPicker(matchId, squads) },
                onFailure = { e ->
                    _step.value = IpscSubmitStep.Error(e.message ?: "Error loading squads") { selectMatch(matchId) }
                }
            )
        }
    }

    fun refreshSquads(matchId: Int) {
        viewModelScope.launch {
            _step.value = IpscSubmitStep.LoadingSquads(matchId)
            repository.getSquadQueue(matchId, forceRefresh = true).fold(
                onSuccess = { squads -> _step.value = IpscSubmitStep.ShooterPicker(matchId, squads) },
                onFailure = { e ->
                    _step.value = IpscSubmitStep.Error(e.message ?: "Error loading squads") { refreshSquads(matchId) }
                }
            )
        }
    }

    fun selectShooter(matchId: Int, shooter: IpscShooter) {
        _step.value = IpscSubmitStep.Confirm(matchId, shooter)
    }

    /**
     * Build and submit an [IpscScoreSubmitRequest] from [summary].
     * Uses [summary.adjustedHitZones] if available, otherwise falls back to
     * [ScoringUtility.calculateEffectiveCounts].
     */
    fun submit(
        matchId: Int,
        shooter: IpscShooter,
        stageId: String,
        summary: DrillRepeatSummary
    ) {
        val hitZones = summary.adjustedHitZones
            ?: ScoringUtility.calculateEffectiveCounts(summary.shots, null)

        val request = IpscScoreSubmitRequest(
            shooterBib = shooter.bibNumber,
            stageId = stageId,
            totalTime = summary.totalTime,
            hits = IpscScoreHits(
                a = hitZones["A"] ?: 0,
                c = hitZones["C"] ?: 0,
                d = hitZones["D"] ?: 0,
                m = hitZones["M"] ?: 0,
                n = hitZones["N"] ?: 0
            ),
            rows = buildRows(summary),
            penalties = IpscScorePenalties(pe = hitZones["PE"] ?: 0),
            firstShot = if (summary.firstShot > 0) summary.firstShot else null,
            fastestSplit = if (summary.fastest > 0) summary.fastest else null
        )

        viewModelScope.launch {
            _step.value = IpscSubmitStep.Submitting
            repository.submitScore(matchId, request).fold(
                onSuccess = { data ->
                    repository.invalidateCache(matchId)
                    _step.value = IpscSubmitStep.Success(data.hitFactor, data.totalPoints)
                },
                onFailure = { e ->
                    _step.value = IpscSubmitStep.Error(e.message ?: "Submission failed") {
                        submit(matchId, shooter, stageId, summary)
                    }
                }
            )
        }
    }

    private data class RowAccumulator(
        val rowType: String,
        val key: String,
        var a: Int = 0,
        var c: Int = 0,
        var d: Int = 0,
        var m: Int = 0,
        var n: Int = 0
    )

    private fun buildRows(summary: DrillRepeatSummary): List<IpscScoreTargetRow>? {
        val grouped = linkedMapOf<String, RowAccumulator>()

        summary.shots.forEach { shot ->
            val targetType = shot.content.actualTargetType.trim().lowercase()
            val rawHitArea = shot.content.actualHitArea
            val isAPopper = isAPopperHitArea(rawHitArea)
            val rowType = if (
                isAPopper ||
                targetType.contains("steel") ||
                targetType.contains("popper") ||
                targetType.contains("paddle")
            ) "steel" else "paper"

            val targetName = shot.target?.trim().takeUnless { it.isNullOrEmpty() }
                ?: shot.content.device?.trim().takeUnless { it.isNullOrEmpty() }
                ?: shot.device?.trim().takeUnless { it.isNullOrEmpty() }
                ?: "target"

            val keyTargetType = if (isAPopper) "apopper" else targetType
            val key = "$rowType|${targetName.lowercase()}|$keyTargetType"
            val row = grouped.getOrPut(key) { RowAccumulator(rowType = rowType, key = key) }

            when (normalizeHitArea(rawHitArea)) {
                "a" -> row.a += 1
                "c" -> row.c += 1
                "d" -> row.d += 1
                "m" -> row.m += 1
                "n" -> row.n += 1
            }
        }

        if (grouped.isEmpty()) return null

        val steelRows = grouped.values
            .filter { it.rowType == "steel" }
            .sortedBy { it.key }
            .mapIndexed { index, row ->
                IpscScoreTargetRow(
                    rowType = "steel",
                    rowNo = index + 1,
                    a = row.a,
                    c = row.c,
                    d = row.d,
                    m = row.m,
                    n = row.n
                )
            }

        val paperRows = grouped.values
            .filter { it.rowType == "paper" }
            .sortedBy { it.key }
            .mapIndexed { index, row ->
                IpscScoreTargetRow(
                    rowType = "paper",
                    rowNo = index + 1,
                    a = row.a,
                    c = row.c,
                    d = row.d,
                    m = row.m,
                    n = row.n
                )
            }

        return steelRows + paperRows
    }

    private fun isAPopperHitArea(raw: String?): Boolean {
        return when (raw?.trim()?.lowercase()) {
            "apopper", "a_popper", "a-popper" -> true
            else -> false
        }
    }

    private fun normalizeHitArea(raw: String?): String {
        return when (raw?.trim()?.lowercase()) {
            "a", "azone", "a_zone", "a-zone", "circlearea", "popperzone", "apopper" -> "a"
            "c", "czone", "c_zone", "c-zone" -> "c"
            "d", "dzone", "d_zone", "d-zone" -> "d"
            "m", "miss" -> "m"
            "n", "ns", "whitezone", "no_shoot", "no-shoot", "noshoot" -> "n"
            "blackzone", "blackzoneleft", "blackzoneright",
            "black_zone", "black-zone", "black_zone_left", "black-zone-left", "black_zone_right", "black-zone-right" -> "m"
            else -> "unknown"
        }
    }

    // ---- Private helpers ----

    private fun loadMatches(forceRefresh: Boolean) {
        viewModelScope.launch {
            _step.value = IpscSubmitStep.LoadingMatches
            repository.getMatches(forceRefresh).fold(
                onSuccess = { matches -> _step.value = IpscSubmitStep.MatchPicker(matches) },
                onFailure = { e ->
                    _step.value = IpscSubmitStep.Error(e.message ?: "Error loading matches") {
                        loadMatches(forceRefresh = true)
                    }
                }
            )
        }
    }
}

File: FlexTargetiOS/flextarget/View/Drills/IpscSubmitViewModel.swift
import Foundation
import Combine

// MARK: - Step

enum IpscSubmitStep: Equatable {
    case idle
    case confirm(context: IpscLockedSelectionContext)
    case submitting
    case success(hitFactor: Double, totalPoints: Int)
    case error(message: String)

    static func == (lhs: IpscSubmitStep, rhs: IpscSubmitStep) -> Bool {
        switch (lhs, rhs) {
        case (.idle, .idle), (.submitting, .submitting):
            return true
        case (.confirm(let a), .confirm(let b)):
            return a.matchId == b.matchId && a.stageId == b.stageId && a.squadId == b.squadId && a.shooter.id == b.shooter.id
        case (.success(let a, let b), .success(let c, let d)):
            return a == c && b == d
        case (.error(let a), .error(let b)):
            return a == b
        default:
            return false
        }
    }
}

// MARK: - ViewModel

@MainActor
final class IpscSubmitViewModel: ObservableObject {

    @Published private(set) var step: IpscSubmitStep = .idle

    private let repository = IpscRepository.shared

    // Summaries provided at the call site (the current drill result)
    var summaries: [DrillRepeatSummary] = []
    var lockedContext: IpscLockedSelectionContext?

    /// Caller-provided row grid (e.g. from the on-screen IPSC editor in
    /// `DrillSummaryView`). When set, this is sent verbatim to the backend so
    /// the payload always carries one row per expected target — including
    /// unengaged targets that default to M=2 — even when no shots were fired.
    var prebuiltRows: [IpscScoreTargetRow]?

    // MARK: - Navigation

    func start() {
        if let lockedContext {
            step = .confirm(context: lockedContext)
            return
        }
        step = .error(message: "Competition context missing. Please start from Competition Session.")
    }

    func dismiss() {
        step = .idle
    }

    func back() {
        switch step {
        case .confirm(let context):
            _ = context
            step = .idle
        case .error:
            step = .idle
        default:
            step = .idle
        }
    }

    /// Build and submit the score.
    func submit(context: IpscLockedSelectionContext, summary: DrillRepeatSummary, isDq: Bool = false) {
        let hitZones = summary.adjustedHitZones
            ?? ScoringUtility.calculateEffectiveCounts(shots: summary.shots, drillSetup: nil)

        let request = IpscScoreSubmitRequest(
            shooterBib: context.shooter.bibNumber,
            stageId: String(context.stageId),
            totalTime: summary.totalTime,
            status: isDq ? .dq : .normal,
            hits: IpscScoreHits(
                A: hitZones["A"] ?? 0,
                C: hitZones["C"] ?? 0,
                D: hitZones["D"] ?? 0,
                M: hitZones["M"] ?? 0,
                N: hitZones["N"] ?? 0
            ),
            rows: prebuiltRows ?? buildRows(from: summary),
            penalties: IpscScorePenalties(PE: hitZones["PE"] ?? 0),
            firstShot: summary.firstShot > 0 ? summary.firstShot : nil,
            fastestSplit: summary.fastest > 0 ? summary.fastest : nil
        )

        step = .submitting
        Task {
            do {
                let data = try await repository.submitScore(matchId: context.matchId, request: request)
                step = .success(hitFactor: data.hitFactor, totalPoints: data.totalPoints)
            } catch {
                if error is DecodingError {
                    step = .error(message: "Server returned incomplete score data. Please retry.")
                } else {
                    step = .error(message: error.localizedDescription)
                }
            }
        }
    }

    private func buildRows(from summary: DrillRepeatSummary) -> [IpscScoreTargetRow]? {
        struct RowAccumulator {
            let rowType: String
            let key: String
            var a: Int = 0
            var c: Int = 0
            var d: Int = 0
            var m: Int = 0
            var n: Int = 0
        }

        var grouped: [String: RowAccumulator] = [:]

        for shot in summary.shots {
            let targetType = shot.content.targetType.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            let rawHitArea = shot.content.hitArea
            let isAPopper = isAPopperHitArea(rawHitArea)
            let rowType = (isAPopper || targetType.contains("steel") || targetType.contains("popper") || targetType.contains("paddle")) ? "steel" : "paper"
            let targetName = (shot.target?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false)
                ? shot.target!.trimmingCharacters(in: .whitespacesAndNewlines)
                : (shot.content.device?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
                    ? shot.content.device!.trimmingCharacters(in: .whitespacesAndNewlines)
                    : (shot.device?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
                        ? shot.device!.trimmingCharacters(in: .whitespacesAndNewlines)
                        : "target"))

            let keyTargetType = isAPopper ? "apopper" : targetType
            let key = "\(rowType)|\(targetName.lowercased())|\(keyTargetType)"
            var row = grouped[key] ?? RowAccumulator(rowType: rowType, key: key)

            switch normalizeHitArea(rawHitArea) {
            case "a": row.a += 1
            case "c": row.c += 1
            case "d": row.d += 1
            case "m": row.m += 1
            case "n": row.n += 1
            default: break
            }

            grouped[key] = row
        }

        if grouped.isEmpty {
            return nil
        }

        let steelRows = grouped.values
            .filter { $0.rowType == "steel" }
            .sorted { $0.key < $1.key }
            .enumerated()
            .map { index, row in
                IpscScoreTargetRow(
                    rowType: "steel",
                    rowNo: index + 1,
                    A: row.a,
                    C: row.c,
                    D: row.d,
                    M: row.m,
                    N: row.n
                )
            }

        let paperRows = grouped.values
            .filter { $0.rowType == "paper" }
            .sorted { $0.key < $1.key }
            .enumerated()
            .map { index, row in
                IpscScoreTargetRow(
                    rowType: "paper",
                    rowNo: index + 1,
                    A: row.a,
                    C: row.c,
                    D: row.d,
                    M: row.m,
                    N: row.n
                )
            }

        return steelRows + paperRows
    }

    private func isAPopperHitArea(_ raw: String) -> Bool {
        return ScoringUtility.normalizeHitArea(raw) == "apopper"
    }

    private func normalizeHitArea(_ raw: String) -> String {
        switch ScoringUtility.normalizeHitArea(raw) {
        case "azone", "circlearea", "popperzone", "apopper":
            return "a"
        case "czone":
            return "c"
        case "dzone":
            return "d"
        case "whitezone":
            return "n"
        case "miss":
            return "m"
        case "n", "ns", "no_shoot", "no-shoot", "noshoot":
            return "n"
        default:
            return "unknown"
        }
    }
}
